File: Tarea8.java
import java.util.ArrayList;
import java.util.Scanner;

public class Tarea8 {

    public static void main(String[] args) {
        Libreria libreria = new Libreria();
        Scanner scanner = new Scanner(System.in);

        while (true) {
            System.out.println("\n Menú:");
            System.out.println("1. Registrar usuario");
            System.out.println("2. Registrar libro");
            System.out.println("3. Rentar libro");
            System.out.println("4. Listar usuarios registrados");
            System.out.println("5. Listar libros registrados");
            System.out.println("6. Listar usuarios con libros rentados");
            System.out.println("7. Listar libros no rentados");
            System.out.println("8. Listar libros rentados");
            System.out.println("9. Salir");
            System.out.print("Seleccione una opción: ");

            String opcion = scanner.nextLine();
            switch (opcion) {
                case "1":
                    System.out.print("Ingrese el nombre del usuario: ");
                    String nombreUsuario = scanner.nextLine();
                    libreria.registrarUsuario(nombreUsuario);
                    break;
                case "2":
                    System.out.print("Ingrese el título del libro: ");
                    String tituloLibro = scanner.nextLine();
                    System.out.print("Ingrese el autor del libro: ");
                    String autorLibro = scanner.nextLine();
                    libreria.registrarLibro(tituloLibro, autorLibro);
                    break;
                case "3":
                    System.out.print("Ingrese el nombre del usuario: ");
                    String nombre = scanner.nextLine();
                    System.out.print("Ingrese el título del libro: ");
                    String titulo = scanner.nextLine();
                    libreria.rentarLibro(nombre, titulo);
                    break;
                case "4":
                    libreria.listarUsuarios();
                    break;
                case "5":
                    libreria.listarLibros();
                    break;
                case "6":
                    libreria.listarUsuariosConLibros();
                    break;
                case "7":
                    libreria.listarLibrosNoRentados();
                    break;
                case "8":
                    libreria.listarLibrosRentados();
                    break;
                case "9":
                    System.exit(0);
                    break;
                default:
                    System.out.println("Seleccione otra opcion");
            }
        }
    }
}

class Libro {
    private String titulo;
    private String autor;
    private boolean rentado;

    public Libro(String titulo, String autor) {
        this.titulo = titulo;
        this.autor = autor;
        this.rentado = false;
    }

    public String obtenerTitulo() {
        return titulo;
    }

    public String obtenerAutor() {
        return autor;
    }

    public boolean estaRentado() {
        return rentado;
    }

    public void establecerRentado(boolean rentado) {
        this.rentado = rentado;
    }
}

class Usuario {
    private String nombre;
    private ArrayList<Libro> librosRentados;

    public Usuario(String nombre) {
        this.nombre = nombre;
        this.librosRentados = new ArrayList<>();
    }

    public void rentarLibro(Libro libro) {
        if (!libro.estaRentado()) {
            librosRentados.add(libro);
            libro.establecerRentado(true);
            System.out.println(nombre + " ha rentado el libro: " + libro.obtenerTitulo());
        } else {
            System.out.println("El libro ya está rentado.");
        }
    }

    public String obtenerNombre() {
        return nombre;
    }

    public ArrayList<Libro> obtenerLibrosRentados() {
        return librosRentados;
    }
}

class Libreria {
    private ArrayList<Usuario> usuarios;
    private ArrayList<Libro> libros;

    public Libreria() {
        usuarios = new ArrayList<>();
        libros = new ArrayList<>();
    }

    public void registrarUsuario(String nombre) {
        usuarios.add(new Usuario(nombre));
        System.out.println("Usuario " + nombre + " registrado");
    }

    public void registrarLibro(String titulo, String autor) {
        libros.add(new Libro(titulo, autor));
        System.out.println("Libro " + titulo + " registrado");
    }

    public void rentarLibro(String nombreUsuario, String tituloLibro) {
        Usuario usuario = null;
        for (Usuario u : usuarios) {
            if (u.obtenerNombre().equals(nombreUsuario)) {
                usuario = u;
                break;
            }
        }
        if (usuario == null) {
            System.out.println("Usuario no encontrado");
            return;
        }

        Libro libro = null;
        for (Libro l : libros) {
            if (l.obtenerTitulo().equals(tituloLibro)) {
                libro = l;
                break;
            }
        }
        if (libro == null) {
            System.out.println("Libro no encontrado");
            return;
        }

        usuario.rentarLibro(libro);
    }

    public void listarUsuarios() {
        System.out.println("Usuarios registrados: ");
        for (Usuario usuario : usuarios) {
            System.out.println(usuario.obtenerNombre());
        }
    }

    public void listarLibros() {
        System.out.println("Libros registrados: ");
        for (Libro libro : libros) {
            System.out.println("Título: " + libro.obtenerTitulo() + ", Autor: " + libro.obtenerAutor());
        }
    }

    public void listarUsuariosConLibros() {
        System.out.println("Usuarios que han rentado al menos un libro:");
        for (Usuario usuario : usuarios) {
            if (!usuario.obtenerLibrosRentados().isEmpty()) {
                System.out.println(usuario.obtenerNombre());
            }
        }
    }

    public void listarLibrosNoRentados() {
        System.out.println("Libros que no han sido rentados:");
        for (Libro libro : libros) {
            if (!libro.estaRentado()) {
                System.out.println("Título: " + libro.obtenerTitulo() + ", Autor: " + libro.obtenerAutor());
            }
        }
    }

    public void listarLibrosRentados() {
        System.out.println("Libros que han sido rentados:");
        for (Libro libro : libros) {
            if (libro.estaRentado()) {
                System.out.println("Título: " + libro.obtenerTitulo() + ", Autor: " + libro.obtenerAutor());
            }
        }
    }
}

File: Tarea8.py
class Libro:
    def __init__(self, titulo, autor):
        self.titulo = titulo
        self.autor = autor
        self.rentado = False

    def obtener_titulo(self):
        return self.titulo

    def obtener_autor(self):
        return self.autor

    def esta_rentado(self):
        return self.rentado

    def establecer_rentado(self, rentado):
        self.rentado = rentado


class Usuario:
    def __init__(self, nombre):
        self.nombre = nombre
        self.libros_rentados = []

    def rentar_libro(self, libro):
        if not libro.esta_rentado():
            self.libros_rentados.append(libro)
            libro.establecer_rentado(True)
            print(self.nombre + " ha rentado el libro: " + libro.obtener_titulo())
        else:
            print("El libro ya está rentado.")

    def obtener_nombre(self):
        return self.nombre

    def obtener_libros_rentados(self):
        return self.libros_rentados


class Libreria:
    def __init__(self):
        self.usuarios = []
        self.libros = []

    def registrar_usuario(self, nombre):
        self.usuarios.append(Usuario(nombre))
        print("Usuario", nombre, "registrado")

    def registrar_libro(self, titulo, autor):
        self.libros.append(Libro(titulo, autor))
        print("Libro", titulo, "registrado")

    def rentar_libro(self, nombre_usuario, titulo_libro):
        usuario = None
        for u in self.usuarios:
            if u.obtener_nombre() == nombre_usuario:
                usuario = u
                break
        if usuario is None:
            print("Usuario no encontrado")
            return

        libro = None
        for l in self.libros:
            if l.obtener_titulo() == titulo_libro:
                libro = l
                break
        if libro is None:
            print("Libro no encontrado")
            return

        usuario.rentar_libro(libro)

    def listar_usuarios(self):
        print("Usuarios registrados:")
        for usuario in self.usuarios:
            print(usuario.obtener_nombre())

    def listar_libros(self):
        print("Libros registrados:")
        for libro in self.libros:
            print("Título:", libro.obtener_titulo(), ", Autor:", libro.obtener_autor())

    def listar_usuarios_con_libros(self):
        print("Usuarios que han rentado al menos un libro:")
        for usuario in self.usuarios:
            if usuario.obtener_libros_rentados():
                print(usuario.obtener_nombre())

    def listar_libros_no_rentados(self):
        print("Libros que no han sido rentados:")
        for libro in self.libros:
            if not libro.esta_rentado():
                print("Título:", libro.obtener_titulo(), ", Autor:", libro.obtener_autor())

    def listar_libros_rentados(self):
        print("Libros que han sido rentados:")
        for libro in self.libros:
            if libro.esta_rentado():
                print("Título:", libro.obtener_titulo(), ", Autor:", libro.obtener_autor())


if __name__ == "__main__":
    libreria = Libreria()

    while True:
        print("\nMenú:")
        print("1. Registrar usuario")
        print("2. Registrar libro")
        print("3. Rentar libro")
        print("4. Listar usuarios registrados")
        print("5. Listar libros registrados")
        print("6. Listar usuarios con libros rentados")
        print("7. Listar libros no rentados")
        print("8. Listar libros rentados")
        print("9. Salir")
        opcion = input("Seleccione una opción: ")

        if opcion == "1":
            nombre_usuario = input("Ingrese el nombre del usuario: ")
            libreria.registrar_usuario(nombre_usuario)
        elif opcion == "2":
            titulo_libro = input("Ingrese el título del libro: ")
            autor_libro = input("Ingrese el autor del libro: ")
            libreria.registrar_libro(titulo_libro, autor_libro)
        elif opcion == "3":
            nombre = input("Ingrese el nombre del usuario: ")
            titulo = input("Ingrese el título del libro: ")
            libreria.rentar_libro(nombre, titulo)
        elif opcion == "4":
            libreria.listar_usuarios()
        elif opcion == "5":
            libreria.listar_libros()
        elif opcion == "6":
            libreria.listar_usuarios_con_libros()
        elif opcion == "7":
            libreria.listar_libros_no_rentados()
        elif opcion == "8":
            libreria.listar_libros_rentados()
        elif opcion == "9":
            break
        else:
            print("Seleccione otra opción")
